# DART 01 - 오픈API 개요

<img src="http://i.imgur.com/W3QMRaX.png" >

* DART 오픈API 사용법
* json_normalize() 활용 
* 기업개황 API, 정보검색 API

### 2018 FinanceData.KR

# 오픈 API
공시 보고서 목록를 실시간 제공
* https://dart.fss.or.kr/dsap001/intro.do 오픈API 소개
* https://dart.fss.or.kr/dsap001/guide.do API 개발 가이드

# API 인증키 발급
계정을 만들고, API 인증키 발급 받는다

계정 신청: https://dart.fss.or.kr/dsag002/insertForm.do

### 개인용,  기업용
* 개인용: 1만건/일
* 기업용: 조회 건수 제한이 없는 대신 조회 서버 IP 제한

# 인증키

* 알파벳과 숫자로 이루어진 40 바이트 문자열
* auth 파라미터로 제공
* ex: 'e81e78aa519d1a1c4ec7867ef12a937ab6cbd4x7'

# 오픈 API가 제공하는 주요정보
주요하게 다음 2가지 API를 제공한다

1. 기업 개황 API : http://dart.fss.or.kr/api/company.json?auth=xxx&crp_cd=xxx
1. 정보 검색 API : http://dart.fss.or.kr/api/search.json?auth=xxx


# 조회 예시

### 요청 URL

```
http://dart.fss.or.kr/api/search.json?auth=e81e78aa519d1a1c4ec7867ef12a937ab6cbd4x7 \
&crp_cd=005930&start_dt=20000101&end_dt=20140730&dsp_tp=A&sort=date
```

### 결과

```javascript
{
  "err_code": "000",
  "err_msg": "정상",
  "page_no": 1,
  "page_set": 10,
  "total_count": 61,
  "total_page": 7,
  "list": [
    {
      "crp_cls": "Y",
      "crp_nm": "삼성전자",
      "crp_cd": "005930",
      "rpt_nm": "[첨부추가]사업보고서 (1999.12)",
      "rcp_no": "20000330000796",
      "flr_nm": "삼성전자",
      "rcp_dt": "20000330",
      "rmk": "연"
    },
    {
      "crp_cls": "Y",
      "crp_nm": "삼성전자",
      "crp_cd": "005930",
      "rpt_nm": "분기보고서 (2000.03)",
      "rcp_no": "20000515000739",
      "flr_nm": "삼성전자",
      "rcp_dt": "20000515",
      "rmk": ""
    }, 
    ...
```

# API 개발 가이드 사이트에서 테스트
https://dart.fss.or.kr/dsap001/guide.do API 개발 가이드

<img src="http://i.imgur.com/9NpoJgx.png" >


# 테스트 인증키
* 테스트 인증키: "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx" 
* HTTP 헤더의 Referer가 "https://dart.fss.or.kr/dsap001/guide.do" 일때 동작

In [1]:
# 실제 인증키를 사용
auth = 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx' 

# 헤더: 실제 인증키를 사용시 사용하지 않아도 된다
headers={'Referer':'https://dart.fss.or.kr/dsap001/guide.do'}

# JSON vs XML
다음 두 가지(XML, JSON) 형식을 제공한다 
* http://dart.fss.or.kr/api/search.xml?auth=xxx (XML)
* http://dart.fss.or.kr/api/search.json?auth=xxx (JSON)

데이터는 동일하다. JOSN이 상대적으로 가볍고, Python에서 다루기도 좋다.

In [2]:
import requests

url_tmpl = 'http://dart.fss.or.kr/api/search.xml?auth={auth}'
url = url_tmpl.format(auth=auth)

r = requests.get(url, headers=headers)
r.text

'<?xml version="1.0" encoding="utf-8"?>\n<result>\n  <err_code>000</err_code>\n  <err_msg>정상</err_msg>\n  <page_no>1</page_no>\n  <page_set>10</page_set>\n  <total_count>29</total_count>\n  <total_page>3</total_page>\n  <list>\n    <crp_cls>K</crp_cls>\n    <crp_nm>알에스오토메이션</crp_nm>\n    <crp_cd>140670</crp_cd>\n    <rpt_nm>임원ㆍ주요주주특정증권등소유상황보고서</rpt_nm>\n    <rcp_no>20180220000014</rcp_no>\n    <flr_nm>박기욱</flr_nm>\n    <rcp_dt>20180220</rcp_dt>\n    <rmk></rmk>\n  </list>\n  <list>\n    <crp_cls>K</crp_cls>\n    <crp_nm>인터불스</crp_nm>\n    <crp_cd>158310</crp_cd>\n    <rpt_nm>조회공시요구(풍문또는보도)(최대주주 지분 매각 추진설)</rpt_nm>\n    <rcp_no>20180220900080</rcp_no>\n    <flr_nm>코스닥시장본부</flr_nm>\n    <rcp_dt>20180220</rcp_dt>\n    <rmk>코</rmk>\n  </list>\n  <list>\n    <crp_cls>K</crp_cls>\n    <crp_nm>디알텍</crp_nm>\n    <crp_cd>214680</crp_cd>\n    <rpt_nm>임원ㆍ주요주주특정증권등소유상황보고서</rpt_nm>\n    <rcp_no>20180220000013</rcp_no>\n    <flr_nm>안홍수</flr_nm>\n    <rcp_dt>20180220</rcp_dt>\n    <rmk></rmk>\n  </li

In [3]:
import requests

url_tmpl = 'http://dart.fss.or.kr/api/search.json?auth={auth}'
url = url_tmpl.format(auth=auth)

r = requests.get(url, headers=headers)
r.text

'{"err_code":"000","err_msg":"정상","page_no":1,"page_set":10,"total_count":29,"total_page":3,"list":[{"crp_cls":"K","crp_nm":"알에스오토메이션","crp_cd":"140670","rpt_nm":"임원ㆍ주요주주특정증권등소유상황보고서","rcp_no":"20180220000014","flr_nm":"박기욱","rcp_dt":"20180220","rmk":""},{"crp_cls":"K","crp_nm":"인터불스","crp_cd":"158310","rpt_nm":"조회공시요구(풍문또는보도)(최대주주 지분 매각 추진설)","rcp_no":"20180220900080","flr_nm":"코스닥시장본부","rcp_dt":"20180220","rmk":"코"},{"crp_cls":"K","crp_nm":"디알텍","crp_cd":"214680","rpt_nm":"임원ㆍ주요주주특정증권등소유상황보고서","rcp_no":"20180220000013","flr_nm":"안홍수","rcp_dt":"20180220","rmk":""},{"crp_cls":"Y","crp_nm":"대교","crp_cd":"019680","rpt_nm":"임원ㆍ주요주주특정증권등소유상황보고서","rcp_no":"20180220000012","flr_nm":"강영중","rcp_dt":"20180220","rmk":""},{"crp_cls":"K","crp_nm":"휴온스","crp_cd":"243070","rpt_nm":"기타경영사항(특허권취득)(자율공시)(RGD모티프 함유 펩티드 또는 이의 단편을 포함하는 화상 및 녹내장 치료, 피부주름 개선 및 발모촉진용 조성물)","rcp_no":"20180220900072","flr_nm":"휴온스","rcp_dt":"20180220","rmk":"코"},{"crp_cls":"K","crp_nm":"디알텍","crp_cd":"214680","rpt_nm":"임원ㆍ주요주주

# 기업개황 API

http://dart.fss.or.kr/api/company.json?auth=xxx&crp_cd=xxx


In [5]:
url_tmpl = 'http://dart.fss.or.kr/api/company.json?auth={auth}&crp_cd={code}'
url = url_tmpl.format(auth=auth, code='005930')
r = requests.get(url, headers=headers)
jo = json.loads(r.text)
jo

{'acc_mt': '12',
 'adr': '경기도 수원시 영통구  삼성로 129 (매탄동)',
 'bsn_no': '1248100998',
 'ceo_nm': '권오현, 윤부근, 신종균',
 'crp_cls': 'Y',
 'crp_nm': '삼성전자(주)',
 'crp_nm_e': 'SAMSUNG ELECTRONICS CO,.LTD',
 'crp_nm_i': '삼성전자',
 'crp_no': '1301110006246',
 'err_code': '000',
 'err_msg': '정상',
 'est_dt': '19690113',
 'fax_no': '031-200-7538',
 'hm_url': 'www.sec.co.kr',
 'ind_cd': '264',
 'ir_url': '',
 'phn_no': '031-200-1114',
 'stock_cd': '005930'}

각 필드의 의미는 다음과 같다

* err_code: 에러코드(오류 메시지 참조)
* err_msg: 에러메시지(오류 메시지 참조)
* crp_nm: 정식명칭
* crp_nm_e: 영문명칭
* crp_nm_i: 종목명(상장사) 또는 약식명칭(기타법인)
* stock_cd: 상장회사인 경우 주식의 종목코드
* ceo_nm: 대표자명
* crp_cls: 법인구분 : Y(유가), K(코스닥), N(코넥스), E(기타)
* crp_no: 법인등록번호
* bsn_no: 사업자등록번호
* adr: 주소
* hm_url: 홈페이지
* ir_url: IR홈페이지
* phn_no: 전화번호
* fax_no: 팩스번호
* ind_cd: 업종코드
* est_dt: 설립일(YYYYMMDD)
* acc_mt: 결산월(MM)

In [6]:
jo['bsn_no'] # 사업자등록번호

'1248100998'

# 정보 검색 API 

http://dart.fss.or.kr/api/search.json?auth={auth}&crp_cd={code}&start_dt=19990101&bsn_tp=A001&bsn_tp=A002&bsn_tp=A003 <br>
(회사의 전체 사업+반기+분기보고서)

* crp_cd: 공시대상회사의 종목코드(상장사:숫자 6자리) 또는 고유번호(기타법인:숫자 8자리)
* end_dt: 검색종료 접수일자(YYYYMMDD) : 없으면 당일
* start_dt: 검색시작 접수일자(YYYYMMDD) : 없으면 end_dt crp_cd가 없는 경우 검색기간은 3개월로 제한
* fin_rpt: 최종보고서만 검색여부(Y or N) 기본값 : N (정정이 있는 경우 최종정정만 검색)
* dsp_tp: 정기공시(A), 주요사항보고(B), 발행공시(C), 지분공시(D), 기타공시(E), 외부감사관련(F), 펀드공시(G), 자산유동화(H), 거래소공시(I), 공정위공시(J)
* bsn_tp: 정기공시(5), 주요사항보고(3), 발행공시(11), 지분공시(4), 기타공시(9), 외부감사관련(3), 펀드공시(3), 자산유동화(6), 거래소공시(6), 공정위공시(5) 
* sort: 접수일자(date), 회사명(crp), 보고서명(rpt) 기본값 : date
* series: 오름차순(asc), 내림차순(desc) 기본값 : desc
* page_no: 페이지 번호(1~n) 기본값 : 1
* page_set: 페이지당 건수(1~100) 기본값 : 10, 최대값 : 100
* callback: 콜백함수명(JSONP용)

# 검색 API 샘플
{auth}는 인증키, {code}는 종목 코드
* 당일 접수 10건 <br> http://dart.fss.or.kr/api/search.json?auth={auth}
* 당일 접수 100건<br> http://dart.fss.or.kr/api/search.json?auth={auth}&page_set=100
* 회사의 당일 접수 10건<br> http://dart.fss.or.kr/api/search.json?auth={auth}&crp_cd={code}
* 회사의 당일 접수 최종보고서만 10건<br> http://dart.fss.or.kr/api/search.json?auth={auth}&crp_cd={code}&fin_rpt=Y
* 회사의 전체(19990101~당일) 공시 10건<br> http://dart.fss.or.kr/api/search.json?auth={auth}&crp_cd={code}&start_dt=19990101
* 회사의 전체 정기공시 10건<br> http://dart.fss.or.kr/api/search.json?auth={auth}&crp_cd={code}&start_dt=19990101&dsp_tp=A
* 회사의 전체 사업+반기+분기보고서 10건<br> http://dart.fss.or.kr/api/search.json?auth={auth}&crp_cd={code}&start_dt=19990101&bsn_tp=A001&bsn_tp=A002&bsn_tp=A003
    

In [7]:
# 당일 접수 10건 

url_tmpl = 'http://dart.fss.or.kr/api/search.json?auth={auth}'
url = url_tmpl.format(auth=auth)
r = requests.get(url, headers=headers)
jo = json.loads(r.text)
result = json_normalize(jo, 'list')
result.head()

crp_cd crp_cls    crp_nm   flr_nm    rcp_dt          rcp_no rmk  \
0  140670       K  알에스오토메이션      박기욱  20180220  20180220000014       
1  158310       K      인터불스  코스닥시장본부  20180220  20180220900080   코   
2  214680       K       디알텍      안홍수  20180220  20180220000013       
3  019680       Y        대교      강영중  20180220  20180220000012       
4  243070       K       휴온스      휴온스  20180220  20180220900072   코   

                                              rpt_nm  
0                                임원ㆍ주요주주특정증권등소유상황보고서  
1                     조회공시요구(풍문또는보도)(최대주주 지분 매각 추진설)  
2                                임원ㆍ주요주주특정증권등소유상황보고서  
3                                임원ㆍ주요주주특정증권등소유상황보고서  
4  기타경영사항(특허권취득)(자율공시)(RGD모티프 함유 펩티드 또는 이의 단편을 포함...

In [8]:
jo

{'err_code': '000',
 'err_msg': '정상',
 'list': [{'crp_cd': '140670',
   'crp_cls': 'K',
   'crp_nm': '알에스오토메이션',
   'flr_nm': '박기욱',
   'rcp_dt': '20180220',
   'rcp_no': '20180220000014',
   'rmk': '',
   'rpt_nm': '임원ㆍ주요주주특정증권등소유상황보고서'},
  {'crp_cd': '158310',
   'crp_cls': 'K',
   'crp_nm': '인터불스',
   'flr_nm': '코스닥시장본부',
   'rcp_dt': '20180220',
   'rcp_no': '20180220900080',
   'rmk': '코',
   'rpt_nm': '조회공시요구(풍문또는보도)(최대주주 지분 매각 추진설)'},
  {'crp_cd': '214680',
   'crp_cls': 'K',
   'crp_nm': '디알텍',
   'flr_nm': '안홍수',
   'rcp_dt': '20180220',
   'rcp_no': '20180220000013',
   'rmk': '',
   'rpt_nm': '임원ㆍ주요주주특정증권등소유상황보고서'},
  {'crp_cd': '019680',
   'crp_cls': 'Y',
   'crp_nm': '대교',
   'flr_nm': '강영중',
   'rcp_dt': '20180220',
   'rcp_no': '20180220000012',
   'rmk': '',
   'rpt_nm': '임원ㆍ주요주주특정증권등소유상황보고서'},
  {'crp_cd': '243070',
   'crp_cls': 'K',
   'crp_nm': '휴온스',
   'flr_nm': '휴온스',
   'rcp_dt': '20180220',
   'rcp_no': '20180220900072',
   'rmk': '코',
   'rpt_nm': '기타경영사항(특허권

# 응답 결과 필드

### result
* err_code: 에러코드(오류 메시지 참조)
* err_msg: 에러메시지(오류 메시지 참조)
* page_no: 페이지 번호
* page_set: 페이지당 건수
* total_count: 총 건수
* total_page: 총 페이지 수

### list
* crp_cls: 법인구분 : Y(유가), K(코스닥), N(코넥스), E(기타)
* crp_nm: 공시대상회사의 종목명(상장사) 또는 법인명(기타법인)
* crp_cd: 공시대상회사의 종목코드(6자리) 또는 고유번호(8자리)
* rpt_nm: 공시구분+보고서명+기타정보
 * [기재정정] : 본 보고서명으로 이미 제출된 보고서의 기재내용이 변경되어 제출된 것임
 * [첨부정정] : 본 보고서명으로 이미 제출된 보고서의 첨부내용이 변경되어 제출된 것임
 * [첨부추가] : 본 보고서명으로 이미 제출된 보고서의 첨부서류가 추가되어 제출된 것임
 * [변경등록] : 본 보고서명으로 이미 제출된 보고서의 유동화계획이 변경되어 제출된 것임
 * [연장결정] : 본 보고서명으로 이미 제출된 보고서의 신탁계약이 연장되어 제출된 것임
 * [발행조건확정] : 본 보고서명으로 이미 제출된 보고서의 유가증권 발행조건이 확정되어 제출된 것임
 * [정정명령부과] : 본 보고서에 대하여 금융감독원이 정정명령을 부과한 것임
 * [정정제출요구] : 본 보고서에 대하여 금융감독원이 정정제출요구을 부과한 것임
* rcp_no: 접수번호(공시뷰어 연결에 이용)
 * PC용 : http://dart.fss.or.kr/dsaf001/main.do?rcpNo=접수번호
 * 모바일용 : http://m.dart.fss.or.kr/html_mdart/MD1007.html?rcpNo=접수번호
* flr_nm: 공시 제출인명
* rcp_dt: 공시 접수일자(YYYYMMDD)
* rmk: 조합된 문자로 각각은 아래와 같은 의미가 있음
 * 유 : 본 공시사항은 한국거래소 유가증권시장본부 소관임
 * 코 : 본 공시사항은 한국거래소 코스닥시장본부 소관임
 * 채 : 본 문서는 한국거래소 채권상장법인 공시사항임
 * 넥 : 본 문서는 한국거래소 코넥스시장 소관임
 * 공 : 본 공시사항은 공정거래위원회 소관임
 * 연 : 본 보고서는 연결부분을 포함한 것임
 * 정 : 본 보고서 제출 후 정정신고가 있으니 관련 보고서를 참조하시기 바람
 * 철 : 본 보고서는 철회(간주)되었으니 관련 철회신고서(철회간주안내)를 참고하시기 바람    

In [9]:
# 당일 접수 100건 (최대 100건)

url_tmpl = 'http://dart.fss.or.kr/api/search.json?auth={auth}&page_set=100'
url = url_tmpl.format(auth=auth)
r = requests.get(url, headers=headers)
jo = json.loads(r.text)
result = json_normalize(jo, 'list')
result.head()

crp_cd crp_cls    crp_nm   flr_nm    rcp_dt          rcp_no rmk  \
0  140670       K  알에스오토메이션      박기욱  20180220  20180220000014       
1  158310       K      인터불스  코스닥시장본부  20180220  20180220900080   코   
2  214680       K       디알텍      안홍수  20180220  20180220000013       
3  019680       Y        대교      강영중  20180220  20180220000012       
4  243070       K       휴온스      휴온스  20180220  20180220900072   코   

                                              rpt_nm  
0                                임원ㆍ주요주주특정증권등소유상황보고서  
1                     조회공시요구(풍문또는보도)(최대주주 지분 매각 추진설)  
2                                임원ㆍ주요주주특정증권등소유상황보고서  
3                                임원ㆍ주요주주특정증권등소유상황보고서  
4  기타경영사항(특허권취득)(자율공시)(RGD모티프 함유 펩티드 또는 이의 단편을 포함...

In [10]:
# 회사의 전체 사업+반기+분기보고서 100건

url_tmpl = 'http://dart.fss.or.kr/api/search.json?auth={auth}&crp_cd={code}&start_dt=19990101&bsn_tp=A001&bsn_tp=A002&bsn_tp=A003&page_set=100'
url = url_tmpl.format(auth=auth, code='005930')
r = requests.get(url, headers=headers)
jo = json.loads(r.text)
result = json_normalize(jo, 'list')
print('총', len(result), '건')
result.head()

총 79 건


crp_cd crp_cls crp_nm flr_nm    rcp_dt          rcp_no rmk           rpt_nm
0  005930       Y   삼성전자   삼성전자  20171114  20171114001694      분기보고서 (2017.09)
1  005930       Y   삼성전자   삼성전자  20170814  20170814001832      반기보고서 (2017.06)
2  005930       Y   삼성전자   삼성전자  20170515  20170515003806      분기보고서 (2017.03)
3  005930       Y   삼성전자   삼성전자  20170331  20170331004518   연  사업보고서 (2016.12)
4  005930       Y   삼성전자   삼성전자  20161114  20161114001491      분기보고서 (2016.09)

# 보고서 보기
조회 결과에서 중요한 것은 문서의 고유 번호인 접수 번호 "rcp_no":"20170515003806" (삼성전자 분기보고서 2017.03)

이 번호를 rcpNo 인자 다음 URL을 만들어 문서를 조회할 수 있다

### 데스크탑 버전
* 'http://dart.fss.or.kr/dsaf001/main.do?rcpNo=' + 접수번호
* http://dart.fss.or.kr/dsaf001/main.do?rcpNo=20170515003806

### 모바일 버전
* 'http://m.dart.fss.or.kr/html_mdart/MD1007.html?rcpNo=' + 접수번호
* http://m.dart.fss.or.kr/html_mdart/MD1007.html?rcpNo=20170515003806

# 보고서 URL 생성

In [11]:
df = result[:10] # 처음 10개만

for ix, row in df.iterrows():
    url_tmpl = "http://dart.fss.or.kr/dsaf001/main.do?rcpNo={}"
    url = url_tmpl.format(row['rcp_no'])
    print (ix, url)

0 http://dart.fss.or.kr/dsaf001/main.do?rcpNo=20171114001694
1 http://dart.fss.or.kr/dsaf001/main.do?rcpNo=20170814001832
2 http://dart.fss.or.kr/dsaf001/main.do?rcpNo=20170515003806
3 http://dart.fss.or.kr/dsaf001/main.do?rcpNo=20170331004518
4 http://dart.fss.or.kr/dsaf001/main.do?rcpNo=20161114001491
5 http://dart.fss.or.kr/dsaf001/main.do?rcpNo=20160816001779
6 http://dart.fss.or.kr/dsaf001/main.do?rcpNo=20160516003174
7 http://dart.fss.or.kr/dsaf001/main.do?rcpNo=20160330003536
8 http://dart.fss.or.kr/dsaf001/main.do?rcpNo=20151116000976
9 http://dart.fss.or.kr/dsaf001/main.do?rcpNo=20150817000859


# Summary
* DART 오픈AP 사용법
* json_normalize() 활용 
* 기업개황 API, 정보검색 API

----
### 2018 FinanceData.KR http://fb.com/financedata